File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.ts
function searchRange(nums: number[], target: number): number[] {
    function findFirst(): number {
        let l = 0, r = nums.length - 1, res = -1;
        while (l <= r) {
            let mid = Math.floor((l + r) / 2);
            if (nums[mid] === target) {
                res = mid;
                r = mid - 1; // search left part
            } else if (nums[mid] < target) {
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        return res;
    }

    function findLast(): number {
        let l = 0, r = nums.length - 1, res = -1;
        while (l <= r) {
            let mid = Math.floor((l + r) / 2);
            if (nums[mid] === target) {
                res = mid;
                l = mid + 1; // search right part
            } else if (nums[mid] < target) {
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        return res;
    }

    const first = findFirst();
    const last = findLast();

    return [first, last];
}

File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int firstOccurance = findFirstOrLastOccurance(nums , target , true);
        int lastOccurance = findFirstOrLastOccurance(nums , target , false);
        return new int[]{firstOccurance,lastOccurance};
    }

    private int findFirstOrLastOccurance(int[] nums , int target, boolean isFindingFirstOccurance){
        int l = 0;
        int r = nums.length - 1;
        int ans = -1;
        while(l <= r) {
            int m = l + (r - l) / 2;
            if(nums[m] < target) l = m + 1;
            else if (nums[m] == target){
                ans = m;
                if(isFindingFirstOccurance) r = m - 1;
                else l = m + 1;
            }
            else r = m - 1;
        }
        return ans;
    }
}